### ISSUE TITLE
runtime error: no attribute 'ipg_index' and 'ipg_buffer' is None

### ISSUE THREAD

**Issue by MannyKayy** [microsoft/DeepSpeed#430]
I am trying to use the cpu-offload feature in a GAN training scenario.

It appears that [this function](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage2.py#L621) can be called when self.ipg_index is not defined simply by enabling the cpu-offload feature in the config file.

**Comment by tjruwase** [comment]
@MannyKayy can you please enable [contiguous gradients](https://www.deepspeed.ai/docs/config-json/#zero-optimizations-for-fp16-training) as well?

**Comment by tjruwase** [comment]
@MannyKayy, actually, please try enabling overlap_comm as well.

**Comment by MannyKayy** [comment]
Thanks, passing in that argument fixes this issue.

**Comment by tjruwase** [comment]
@MannyKayy Thanks for this update. But I am a bit confused by this issue because of this [line](https://github.com/microsoft/DeepSpeed/blob/f0f2a7026877d3fd2f6f5565a67cdffc89750cf0/deepspeed/runtime/zero/stage2.py#L303). 

As you can see, enabling cpu_offload should automatically enable contiguous gradients. Can you confirm that you saw this issue with the latest code?

**Comment by MannyKayy** [comment]
yep, I am using the latest codebase

**Comment by MannyKayy** [comment]
@tjruwase Sorry, correction.. I'm using [this revision](https://github.com/microsoft/DeepSpeed/commit/473ff985fb8770fbbd9073554a666aac683d57d2). I'll pull the current repo and see if the problem persists.

**Comment by tjruwase** [comment]
@MannyKayy, no worries. It is possible we have a regression, so please share your findings. Thanks!

**Comment by MannyKayy** [comment]
The latest revision no longer raises this issue. Just enabling cpu_offload works, even without enabling contiguous gradients. Thanks once again.

**Comment by MannyKayy** [comment]
Okay, so it appears that this issue remains even in the latest version of the repo... 

Also correcting my earlier comment; as long as "DeepSpeedCPUAdam" is in use; this issue appears regardless of whether cpu_offload, overlap_comm and contiguous gradients are all enabled...

Deepspeed Config File:
`{
  "train_micro_batch_size_per_gpu": 64,
  "gradient_accumulation_steps": 2,
  "steps_per_print": 100,
  "zero_optimization": {
    "stage": 2,
    "cpu_offload": true,
    "overlap_comm": true,
    "contiguous_gradients": true,
    "reduce_bucket_size": 50000000
  },
  "zero_allow_untested_optimizer": true,
  "gradient_clipping": 1.0,
  "fp16": {
    "enabled": true,
    "loss_scale": 4096,
    "loss_scale_window": 1000,
    "hysteresis": 2,
    "min_loss_scale": 1
  },
  "activation_checkpointing": {
    "partition_activations": true,
    "contiguous_memory_optimization": true,
    "cpu_checkpointing": true
  },
  "wall_clock_breakdown": false
}
`

**Comment by MannyKayy** [comment]
So after abit of debugging, it appears that this issue is a version of this [one](https://github.com/microsoft/DeepSpeed/issues/399). 

In a multi-engine, multi-optimizer setup; if at any point you forget to freeze the state for the parameters being handled by other optimizers during an update.. it raises a set of issues such as this one..

I simple solution in the gan scenario is to make sure that grads are only enabled for the model being updated.

```
def toggle_grad(model, require_grads=False):
  for param in model.parameters():
  param.requires_grad = require_grads
```

**Comment by SophieOstmeier** [comment]
This issue still occurs. This https://deepspeed.readthedocs.io/en/latest/_modules/deepspeed/runtime/zero/stage_1_and_2.html has no attribute ipg_index, when training with optimizer stage 2. 
Is there any fix to this?

**Comment by GaParmar** [comment]
I also ran into this issue, the solution posted by @MannyKayy worked for me!
